Refactor event registry

bl4ckscor3 opened the pull request:
### Motivation
The previous system had several flaws, mainly the fact that event IDs were tied to class names. With the port to 1.20.6, custom event data also became less flexible.

### Implementation
Suppliers for event constructors are no longer registered directly. Instead, they are wrapped in an `EventType`, which is now registered to a `Registry`, to have a proper ID like blocks etc.. This `EventType` also holds additional data, some of which was previously configured in the `Event` class and has been moved to here.
An `EventType` consists of:
- A `Supplier` for creating a new event instance
- A `StreamCodec` for sending the event to the client
- `requiredFeatures`, which was previously a separate map, holding the set of feature flags that needs to be enabled in order for the event to work
- `disabledByAccessibilityMode` for whether accessibility mode disables this event
- `category`, previously implemented by `Event#type` [^1]

When an event is selected, it will be wrapped in a new `TypedEvent` struct. It serves to hold the event instance, as well as its associated `EventType`. This struct is sent to connected/-ing clients - instead of previously the event ID - and will also encode/decode special event data as dictated by the `EventType#streamCodec`. The latter is currently only used by the fake teleport events.

Alongside this change, I renamed all event IDs to snake_case and adjusted translation keys to be based on the registration IDs.
The config for disabled events was also renamed to prevent crashes from a now malformed config due to changed event IDs.

Minor changes in this PR include fixing of a few logging message parameters, and removal of unused fields & methods.

[^1]: This can likely still be improved, but in order to keep the scope small within an already large PR, I opted to not refactor this, too

slicedlime commented:
Some initial reflections after looking at this, but not digging through and doing a full review:

- Using a Registry is spot on to me - much of this stuff seems like the right way to go
- I don't understand the TypedEvent structure - I would've expected a `getType()` on event, which would enable dispatch through the registry, which is the built-in way to do all the boilerplate you have for the encoding of event data
- Doing that would also enable using codecs (non-stream) to store events, which is another thing I'd like to do to not have the whole mod seemingly start over when you save and load back in
- I'd suggest taking the opportunity to make the category into either an enum or another registry type, since it'd be nice to have a list to ensure we have translations and can show categories

juancarloscp52 commented:
> * I don't understand the TypedEvent structure - I would've expected a `getType()` on event, which would enable dispatch through the registry, which is the built-in way to do all the boilerplate you have for the encoding of event data

I agree with this, seems this would avoid a bit of complexity instead of using TypedEvent.

> * I'd suggest taking the opportunity to make the category into either an enum or another registry type, since it'd be nice to have a list to ensure we have translations and can show categories

 I'm very interested on exploring the possibility of showing categories, specially for filtering the event list. (and any other ideas you might have) I agree with changing this to an Enum. (in this PR or on a later one, no hurry)


With that said, overall the PR looks good and a great idea, although I did not review it fully atm.

slicedlime commented:
I don't think you need to convert away from an interface at all. All you need is an `EventType<? extends Event> type();` method in the interface - vanilla has plenty of examples of this - look at EnchantmentProvider for a simple example (although it's so simple that the type in that case is just a codec)

bl4ckscor3 commented:
Removed `TypedEvent`, and changed category from a string to a new `EventCategory` enum. However as said before, I did not implement a list of categories or any user-facing features regarding them.

slicedlime commented on src/main/java/me/juancarloscp52/entropy/server/ServerEventHandler.java:
```diff
@@ -88,13 +89,13 @@ public void tick(boolean noNewEvents) {
                         else    // Get winner
                             event = voting.events.get(winner);
-                        Entropy.LOGGER.info("[Chat Integrations] Winner event: " + EventRegistry.getTranslationKey(event));
+                        Entropy.LOGGER.info("[Chat Integrations] Winner event: {}", EventRegistry.getTranslationKey(event.getType()));
```
Pre-existing, but why translation key raw output into log?

slicedlime reviewed (changes requested):
Partial review of everything except the event DB for now.

bl4ckscor3 commented on src/main/java/me/juancarloscp52/entropy/events/Event.java:
```diff
@@ -47,7 +59,7 @@ default void render(GuiGraphics drawContext, DeltaTracker tickCounter) {
 
     @Environment(EnvType.CLIENT)
-    void renderQueueItem(GuiGraphics drawContext, float tickdelta, int x, int y);
+    void renderQueueItem(EventType<?> eventType, GuiGraphics drawContext, float tickdelta, int x, int y);
```
Added to allow for changing which event is being displayed (used by fake (fake) teleport events). Probably not the best solution.

slicedlime replied:
Ah, I see. I'd suggest moving the translation keys into EventType and having a `getName()` or something on the event then - so the fake events could override that.

slicedlime commented on src/main/java/me/juancarloscp52/entropy/events/EventRegistry.java:
```diff
@@ -17,273 +17,279 @@
+        register(registry, "rainbow_trails", RainbowTrailsEvent.TYPE);
+        register(registry, "rainbow_sheep_everywhere", RainbowSheepEverywhereEvent.TYPE);
+        register(registry, "armor_trim", ArmorTrimEvent.TYPE);
+        return registry.freeze();
```
Anything that requires lookup from the root registry. It's possible that the state on your branch works without it, but many of the suggested changes will crash without it. You've essentially paid for the functionality in re-implementing the built-in functionality that needs lookups.

bl4ckscor3 commented on src/main/java/me/juancarloscp52/entropy/events/Event.java:
```diff
@@ -21,11 +21,23 @@
 
 
 public interface Event {
+    static <T extends Event> StreamCodec<FriendlyByteBuf, T> streamCodec(EventType.EventSupplier<T> eventSupplier) {
+        return new StreamCodec<>() {
+            @Override
+            public T decode(FriendlyByteBuf buf) {
+                return eventSupplier.create();
+            }
+
+            @Override
+            public void encode(FriendlyByteBuf buf, T event) {}
+        };
+    }
```
The constructor isn't networked though? This is essentially a no-op stream codec that just invokes the constructor. The server has no data to send to the client, so nothing is written to the buffer. The client reads nothing from it either. I think events on the client should be created with the data the server sends, so for events with no data there is nothing else to call but the empty constructor. The client uses the stream codec to create events with data (see fake teleport events), so how else can this be done?

slicedlime replied:
Ah, I see - I think I misinterpreted it because of the generic name. That sounds like `StreamCodec.unit()` though, so this shouldn't be needed.

bl4ckscor3 replied:
Derp.

bl4ckscor3 replied:
Ah, but `StreamCodec.unit()` requires the same object, while events need to be recreated each time they're run.

slicedlime replied:
Yeah, it'd require either making them singletons or records. Either would be fine, but seems not worthwhile for that purpose.

The current implementation is surprising to me, because it sits in a builder and could apply to any type, even the ones with data. Can leave that for later when we had time to think up some nice way of making it clearer though.

bl4ckscor3 commented on src/main/java/me/juancarloscp52/entropy/events/EventRegistry.java:
```diff
@@ -17,273 +17,279 @@
-    public static HashMap<String, FeatureFlagSet> requiredFeatures;
+    public static final ResourceKey<Registry<EventType<?>>> REGISTRY_KEY = ResourceKey.createRegistryKey(ResourceLocation.fromNamespaceAndPath("entropy", "events"));
+    public static final Registry<EventType<?>> EVENTS = bootstrap();
+    public static final StreamCodec<FriendlyByteBuf, Event> STREAM_CODEC = new StreamCodec<>() {
```
Moving to Event interface/leaving it below the registry seems to crash the game whenever an event that uses that stream codec is run.

bl4ckscor3 replied:
I ran the fake teleport event through `/entropy run entropy:fake_teleport` with the `STREAM_CODEC` field as it is now, moved to the `Event` interface. It could be that the problem arises from the fake teleport event's stream codec referencing that base event stream codec, which then causes an NPE. The stacktrace is a bit tricky to sift through, but using the base event stream codec as I had it originally before addressing this simplification makes the error a bit more clear.

slicedlime replied:
Not worth stalling this PR for anyway - can figure that out later.

slicedlime commented on src/main/java/me/juancarloscp52/entropy/events/db/LowFPSEvent.java:
```diff
@@ -41,12 +42,7 @@ public void endClient() {
     }
 
     @Override
-    public String type() {
-        return "fps";
-    }
```
This disappeared. On purpose?

bl4ckscor3 replied:
Yes, this was the only event that had this type.

slicedlime commented on src/main/java/me/juancarloscp52/entropy/events/db/DVDEvent.java:
```diff
@@ -18,15 +18,19 @@
 
 public class DVDEvent extends AbstractTimedEvent {
 
+    public static final EventType<DVDEvent> TYPE = EventType.builder(DVDEvent::new).category(EventCategory.SCREEN_ASPECT).build();
```
Change in category here? That change makes sense to me, but wanted to make sure it's on purpose.

bl4ckscor3 replied:
It's on purpose, as it's the only event that had that type which is now no longer the case being properly categorized.

slicedlime reviewed (approved):
Rest of the review.

Good work!

bl4ckscor3 commented on src/main/java/me/juancarloscp52/entropy/events/EventRegistry.java:
```diff
@@ -17,273 +17,279 @@
-        String[] name = event.getClass().getName().split("\\.");
-        return name[name.length - 1];
+    public static ResourceLocation getEventId(EventType<?> eventType) {
+        return EVENTS.getKey(eventType);
```
Guessing this can be resolved since the method changed to getting the ResourceKey now.
